packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,245 @@
+   * @param options.messenger - The controller messenger.
+   * @param options.isEnabled - Function that returns whether the controller is enabled. (default: () => true)
+   */
+  constructor({
+    messenger,
+    isEnabled = () => true,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    isEnabled?: () => boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      messenger,
+      state: getDefaultDefiPositionsControllerState(),
+    });
+
+    this.setIntervalLength(TEN_MINUTES_IN_MS);
+
+    this.#fetchPositions = buildPositionFetcher();
+    this.#isEnabled = isEnabled;
+
+    this.messagingSystem.subscribe('KeyringController:unlock', async () => {
+      this.startPolling(null);
+    });
+
+    this.messagingSystem.subscribe('KeyringController:lock', () => {
+      this.stopAllPolling();
+    });
+
+    this.messagingSystem.subscribe(
+      'TransactionController:transactionConfirmed',
+      async (transactionMeta) => {
+        if (!this.#isEnabled()) {
+          return;
+        }
+
+        await this.#updateAccountPositions(transactionMeta.txParams.from);
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => {
+        if (!this.#isEnabled() || !account.type.startsWith('eip155:')) {
+          return;
+        }
+
+        await this.#updateAccountPositions(account.address);
+      },
+    );
+  }
+
+  async _executePoll(): Promise<void> {
+    if (!this.#isEnabled()) {
+      return;
+    }
+
+    const accounts = this.messagingSystem.call(
+      'AccountsController:listAccounts',
+    );
+
+    const results = await reduceInBatchesSerially({
+      initialResult: [] as {
+        accountAddress: string;
+        positions: GroupedPositionsPerChain | null;
+      }[],
+      values: accounts,
+      batchSize: FETCH_POSITIONS_BATCH_SIZE,
+      eachBatch: async (workingResult, batch) => {
+        const batchResults = (
+          await Promise.all(
+            batch.map(async ({ address: accountAddress, type }) => {
+              if (type.startsWith('eip155:')) {
+                const positions =
+                  await this.#fetchAccountPositions(accountAddress);
+
+                return {
+                  accountAddress,
+                  positions,
+                };
+              }
+
+              return undefined;
+            }),
+          )
+        ).filter(Boolean) as {
+          accountAddress: string;
+          positions: GroupedPositionsPerChain | null;
+        }[];
+
+        return [...workingResult, ...batchResults];
+      },
+    });
+
+    const allDefiPositions = results.reduce(
+      (acc, { accountAddress, positions }) => {
+        acc[accountAddress] = positions;
+        return acc;
+      },
+      {} as DeFiPositionsControllerState['allDeFiPositions'],
+    );
+
+    this.update((state) => {
+      state.allDeFiPositions = allDefiPositions;
```

Comment:
The batching logic above is relatively new to me. Does it continue with next batch and we filter out batches that fail?

Reason I'm asking, is here we overwrite the defi positions. Is there a possibility of us overwriting some positions for an address that didn't resolve?